### PR TITLE
add typescript to externally facing files

### PR DESCRIPTION
**Pull request by obmelvin**

### CODE REVIEW

**Comment by obmelvin** [comment]
Do we want to upgrade eslint to major version 7 to support the rule that is causing the circleci test to fail?

File: .eslintrc
Comment (obmelvin) [inline .eslintrc:2]: see comments regarding ts parser vs babel parser

File: package.json
Comment (obmelvin) [inline package.json:35]: we should only keep this or @babel/eslint-parser in the end

File: src/form/Form.d.ts
Comment (obmelvin) [inline src/form/Form.d.ts:65]: not sure about this type

File: src/utils/init.d.ts
Comment (obmelvin) [inline src/utils/init.d.ts:11]: another question

File: tsconfig.json
Comment (obmelvin) [inline tsconfig.json:1]: I left all these options in w/ the provided explanations so that people could review them. Will cut down to only those in use before merging

File: package.json
Comment (obmelvin) [inline package.json:8]: dropping the .js extension here was crucial for this all to work

**Comment by bo-dun-1** [comment]
> Do we want to upgrade eslint to major version 7 to support the rule that is causing the circleci test to fail?

upgrading seems reasonable to me. will that require a refactor first?

File: src/form/Form.d.ts
Comment (bo-dun-1) [inline src/form/Form.d.ts:12]: the onChange trigger can only be field or addressSelect currently while the onSubmit and onSkip can be click or change. this is all in our documentation https://docs.feathery.io/develop/react/onsubmit

File: src/form/Form.d.ts
Comment (bo-dun-1) [reply src/form/Form.d.ts:65]: yeah plaid fields are a slightly different structure since they're hidden fields. normal formatted field payloads are { displayText, value, type} while hidden field payload is just value. wonder if we should maybe structure it so hidden fields are more consistent with form fields? or shoudl we just update the typings

File: tsconfig.json
Comment (bo-dun-1) [reply tsconfig.json:1]: will defer to @drewj @ericpias @ondrosh here

File: tsconfig.json
Comment (ericpias) [inline tsconfig.json:11]: We don't need it now I guess, but options like allowJs and checkJs will likely be needed later if we begin introducing typescript into the lib.

File: tsconfig.json
Comment (ericpias) [reply tsconfig.json:1]: The enabled options seem reasonable to me when just adding typings.

File: src/utils/init.d.ts
Comment (ericpias) [inline src/utils/init.d.ts:21]: This may be slightly pedantic, but the docs use the term 'fieldKey' instead of 'fieldName'.  We should probably keep in consistent.

File: src/utils/init.d.ts
Comment (ericpias) [inline src/utils/init.d.ts:16]: validateStep has return type {<fieldKey>: <errorMessage>}

File: src/elements/index.d.ts
Comment (ericpias) [inline src/elements/index.d.ts:1]: I guess just using an untyped intrinsic here?  Seems fine for now but we will want to type deeper if the dashboard starts converting to typescript.

File: src/utils/init.d.ts
Comment (ericpias) [reply src/utils/init.d.ts:11]: Void seems right to me.  The docs make no "promises" on nor mention of the return type.

File: src/form/Form.d.ts
Comment (ericpias) [inline src/form/Form.d.ts:6]: The property can also be an Object type (specifically documented for a style but possibly other properties).  Not sure how deep this could go.

File: src/form/Form.d.ts
Comment (ericpias) [inline src/form/Form.d.ts:23]: Yes.  I think a type enum makes sense.

File: src/form/Form.d.ts
Comment (ericpias) [inline src/form/Form.d.ts:25]: Should we just punt on this and use any here instead?

**Review by ericpias — COMMENTED** [review]
I would also suggest that we test with a react TS app embedding a form if you haven't already.  Also, it might be good to test it in a NextJS typescript app as well since we probably know that the first customer use case is there.

File: src/form/Form.d.ts
Comment (ericpias) [inline src/form/Form.d.ts:26]: Should this be named 'FieldData' to better match the docs?

**Comment by obmelvin** [comment]
> I would also suggest that we test with a react TS app embedding a form if you haven't already. Also, it might be good to test it in a NextJS typescript app as well since we probably know that the first customer use case is there.

I have tested it with local hosted forms for a couple of forms, but I haven't tried with a nextJS app. That would definitely be worthwile

File: src/utils/init.d.ts
Comment (bo-dun-1) [reply src/utils/init.d.ts:11]: if you look at the init definition, it returns an awaitable promise that returns void as you specified here. if the docs don't mention it could you go in and add it?

Comment (bo-dun-1) [reply src/utils/init.d.ts:11]: specifying the awaitability here IMO is important

File: src/index.d.ts
Comment (bo-dun-1) [inline src/index.d.ts:24]: have you tried linking this to the dashboard to test form elements?

**Review by bo-dun-1 — COMMENTED** [review]
i tested some of the callback functions, the children component, and getAllValues in hosted forms. all seemed to work well for me!

File: src/utils/init.d.ts
Comment (obmelvin) [reply src/utils/init.d.ts]: We do say that it is an async function function in the docs

File: src/elements/index.d.ts
Comment (obmelvin) [reply src/elements/index.d.ts]: @ericpias in the new src/elements/index.ts file you can see that the Elements type is inferred from all the elements spread into the object

File: .eslintrc
Comment (bo-dun-1) [inline .eslintrc:34]: make this a TODO

**Review by bo-dun-1 — APPROVED** [review]

**Pull request merged**
